BEIMA.Backend/ReportService/ReportWriter.cs

```diff
@@ -0,0 +1,275 @@
+                        {
+                            continue;
+                        }
+
+                        var deviceTypeDevices = typeToDevices[deviceType.Id];
+
+                        var fileInZip = zipStream.CreateEntry($"{deviceType.Name}.csv", CompressionLevel.Optimal); // Create new zip file entry that can be streamed into
+                        using (var fileInZipStream = fileInZip.Open())                                             // Open zip file entry's stream
+                        using (var dataStream = new MemoryStream())                                                // Create memory stream that data can be written into
+                        using (var csvWriter = new StreamWriter(dataStream))                                       // Create writer to write device/device type data to memory stream
+                        {
+                            
+
+                            // Write file headers                            
+                            var headers = GenerateDeviceReportHeaders(deviceType);
+                            var headerString = string.Join(delimiter, headers);
+                            if (deviceTypeDevices.Count > 0)
+                            {
+                                headerString += Environment.NewLine;
+                            }
+                            csvWriter.Write(headerString);
+
+                            // Write all device data                            
+                            for (var i = 0; i < deviceTypeDevices.Count; i++)
+                            {
+                                // Write device's values
+                                var device = deviceTypeDevices[i];
+                                var values = GenerateDeviceValues(device, deviceType);
+                                var valueString = string.Join(delimiter, values);
+                                if(i != deviceTypeDevices.Count - 1)
+                                {
+                                    valueString += Environment.NewLine;
+                                }
+                                csvWriter.Write(valueString);
+                            }
+                            // Flush writer and copy data in datastream to zip file entry's stream
+                            csvWriter.Flush();
+                            dataStream.Position = 0;
+                            dataStream.CopyTo(fileInZipStream);
+                        }
+                    }
+                }
+                zipFileBytes = outStream.ToArray();
+            }
+            return zipFileBytes;
+        }
+
+        /// <summary>
+        /// Creates a list of strings filled with all of the property names in the Device object. The Fields
+        /// property is replaced with the field names contained in the parameter deviceType's Fields dictionary
+        /// </summary>
+        /// <param name="deviceType"></param>
```

Comment:
Missed a couple params here and in IMongoConnector

Reply:
Should be there now